sinks/grpsink/grpc_stream.go

```diff
@@ -0,0 +1,273 @@
+	name = "grpc-" + name
+	// We want the stream in fail-fast mode. This is the default, but it's
+	// important to the design, so it's worth being explicit: if the streams
+	// weren't in fail-fast mode, then Ingest() calls will block while connections
+	// get re-established in the background, resulting in undesirable backpressure.
+	// For a use case like this sink, unequivocally preferred to fail fast instead,
+	// resulting in dropped spans.
+	dco := grpc.WithDefaultCallOptions(grpc.FailFast(true))
+	conn, err := grpc.DialContext(ctx, target, append(opts, dco)...)
+	if err != nil {
+		log.WithError(err).WithFields(logrus.Fields{
+			"name":   name,
+			"target": target,
+		}).Error("Error establishing connection to gRPC server")
+		return nil, err
+	}
+
+	return &GRPCStreamingSpanSink{
+		grpcConn:   conn,
+		name:       name,
+		target:     target,
+		sinkClient: NewSpanSinkClient(conn),
+		commonTags: commonTags,
+		log:        log,
+	}, nil
+}
+
+// Start performs final preparations on the sink before it is
+// ready to begin ingesting spans.
+func (gs *GRPCStreamingSpanSink) Start(cl *trace.Client) error {
+	gs.traceClient = cl
+
+	var err error
+	gs.stream, err = gs.sinkClient.SendSpans(context.TODO())
+	if err != nil {
+		gs.log.WithFields(logrus.Fields{
+			"name":          gs.name,
+			"target":        gs.target,
+			"chanstate":     gs.grpcConn.GetState().String(),
+			logrus.ErrorKey: err,
+		}).Error("Failed to set up a stream over gRPC channel to sink target")
+		return err
+	}
+
+	go gs.maintainStream()
+	return nil
+}
+
+// maintainStream is intended to be run in a background goroutine. It is kicked
+// off by Start(), and is only kept as a standalone method for testing purposes.
+//
+// This method runs an endless loop that reacts to state changes in the
+// underlying channel by automatically attempting to re-establish the stream
+// connection when it moves back into the 'READY' state. See
+// https://github.com/grpc/grpc/blob/master/doc/connectivity-semantics-and-api.mdA
+// for details about gRPC's connectivity state machine.
+func (gs *GRPCStreamingSpanSink) maintainStream() {
+	for {
+		// This call will block on a channel receive until the gRPC connection
+		// becomes unhealthy. Then, spring into action!
+		state := gs.grpcConn.GetState()
+		switch state {
+		case connectivity.Idle:
+			// gRPC considers an open stream to be an active RPC, and channels
+			// normally only go idle if there are no active RPC within the
+			// channel's timeout window. It should be nigh-impossible for that
+			// to happen here, so we record entering the Idle state as an error.
+			gs.log.WithFields(logrus.Fields{
+				"name":          gs.name,
+				"target":        gs.target,
+				"chanstate":     gs.grpcConn.GetState().String(),
+				logrus.ErrorKey: fmt.Errorf("gRPC sink became idle; should be nearly impossible"),
+			}).Error("gRPC channel transitioned to idle")
```

Comment:
this will record a sentry error if a veneur should not receive spans... I guess that's nearly impossible, but still seems a bit excessive? Maybe we can just increment a metric instead?

Reply:
i chose error for this because we consider it an unreachable state. So unreachable that i would ordinarily `panic()` here, as a canary indicator that there's a bug in the code/our assumptions. IOW, yeah, i'd hope to see a sentry error, because if these assumptions are wrong, then we need to revisit some basic thinking about the sink :)